src/economy.c

```diff
@@ -299,6 +316,120 @@ void commodity_Jettison( int pilot, Commodity* com, int quantity )
 }
 
 
+/**
+ * @brief Initializes a gatherable object
+ *
+ *    @param com Type of commodity.
+ *    @param pos Position.
+ *    @param vel Velocity.
+ */
+void gatherable_init( Commodity* com, Vector2d pos, Vector2d vel )
+{
+   gatherable_stack = realloc(gatherable_stack,
+                              sizeof(Gatherable)*(++gatherable_nstack));
+
+   gatherable_stack[gatherable_nstack-1].type = com;
+   gatherable_stack[gatherable_nstack-1].pos = pos;
+   gatherable_stack[gatherable_nstack-1].vel = vel;
+   gatherable_stack[gatherable_nstack-1].timer = 0.;
+   gatherable_stack[gatherable_nstack-1].lifeleng = RNGF()*100. + 50.;
+}
+
+
+/**
+ * @brief Updates all gatherable objects
+ *
+ *    @param dt Elapsed time.
+ */
+void gatherable_update( double dt )
+{
+   int i;
+
+   for (i=0; i < gatherable_nstack; i++) {
+      gatherable_stack[i].timer += dt;
+      gatherable_stack[i].pos.x += dt*gatherable_stack[i].vel.x;
+      gatherable_stack[i].pos.y += dt*gatherable_stack[i].vel.y;
+
+      /* Remove the gatherable */
+      if (gatherable_stack[i].timer > gatherable_stack[i].lifeleng) {
+         gatherable_nstack--;
+         memmove( &gatherable_stack[i], &gatherable_stack[i+1],
+                 sizeof(Gatherable)*(gatherable_nstack-i) );
+         gatherable_stack = realloc(gatherable_stack,
+                                    sizeof(Gatherable) * gatherable_nstack);
+         i--;
+      }
+   }
+}
+
+
+/**
+ * @brief Frees all the gatherables
+ */
+void gatherable_free( void )
+{
+   free(gatherable_stack);
+   gatherable_stack = NULL;
+   gatherable_nstack = 0;
+}
+
+
+/**
+ * @brief Renders all the gatherables
+ */
+void gatherable_render( void )
+{
+   int i;
+   Gatherable *gat;
+
+   for (i=0; i < gatherable_nstack; i++) {
+      gat = &gatherable_stack[i];
+      gl_blitSprite( gat->type->gfx_space, gat->pos.x, gat->pos.y, 0, 0, NULL );
+   }
+}
+
+
+/**
+ * @brief See if the pilot can gather anything
+ *
+ *    @param pilot ID of the pilot
+ */
+void gatherable_gather( int pilot )
+{
+   int i, q;
+   Gatherable *gat;
+   Pilot* p;
+
+   p = pilot_get( pilot );
+
+   for (i=0; i < gatherable_nstack; i++) {
+      gat = &gatherable_stack[i];
+
+      if ( .03*vect_dist( &p->solid->pos, &gat->pos ) +
+           .03*vect_dist( &p->solid->vel, &gat->vel )  < 1. ) {
+         /* Add cargo to pilot. */
+         q = pilot_cargoAdd( p, gat->type, RNG(1,5), 0 );
```

Comment:
This can fail if the ship does not have cargo space left. It may be worth displaying a message in this case; otherwise it is confusing when nothing is gathered.